### PR TITLE
Narcissus did not accept keywords as property names

### PR DESCRIPTION
**Pull request by bjouhier**
Narcissus only accepted identifiers as property names. From the ES5 specs, it seems that keywords are also allowed (IdentifierName vs. Identifier).

Bruno

### CODE REVIEW

File: lib/jslex.js [hunk 1]
```diff
@@ -107,6 +107,19 @@ Narcissus.lexer = (function() {
             }
             return this.token;
         },
+        
+        forceIdentifier: function() {
+        	if (!this.match(IDENTIFIER)) {
+        		// keywords are valid property names in ES 5
+        		if (this.get() >= definitions.keywords[0] || this.unget) {
```
Comment (syg) [inline lib/jslex.js:114]: I feel like doing `this.get() >= definitions.keywords[0]` is too hacky and depends on keywords being at the end of the tokens table. I'd rather that the actual `definitions.keywords` map be consulted.

Also, `this.unget` needs to be `this.unget()`.

Comment (bjouhier) [inline lib/jslex.js:114]: No problem with me. This was a just quick fix to get going, there is certainly a better way to do it.
Bruno

**Comment by bjouhier** [comment]
There was another bug with property names. Property names starting a digit (`var foo = { 10k: "baz" }` for ex) were not quoted.

**Comment by syg** [comment]
Property names that start with a digit but which are not numeric literals are not supposed to be automatically quoted, I don't think.

From ES5 11.1.5:

PropertyName ::= IdentifierName | StringLiteral | NumericLiteral

From ES5 7.6:

IdentifierName cannot start with a digit (ES5 7.6).

**Comment by bjouhier** [comment]
The problem that I have is the following:
- Narcissus parses `var foo = { "10k": "baz" }` without problem
- But when I regenerate the source with the decompiler, I get `var foo = { 10k: "baz" }`

The quotes are gone. This is not a parser problem, but a decompiler problem: the regenerated code is invalid.

**Comment by syg** [comment]
Oh! Yes, that's a problem.

**Comment by bjouhier** [comment]
And the fix is easy! Just one test to change.

**Comment by syg** [comment]
Could you change the pull request to only include be83c4e? The previous patch no longer applies and the bug you reported should be fixed in master.

**Comment by bjouhier** [comment]
OK.

2011/8/27 syg <
reply@reply.github.com>

**Comment by bjouhier** [comment]
I'm having difficulties with my internet connection. Will retry a bit later. Sorry

Envoyé de mon iPhone

**Comment by bjouhier** [comment]
Looks like my internet connection is working again. 

I did not find any easy way to delete the commit from the pull request. Can you just pick the one line change from be83c4e and apply it.

I will then close the pull request. This way, if I have another fix to submit later, I'll do it as a separate pull request.

Bruno

**Comment by syg** [comment]
That sounds fine. I have to run for now, but I'll commit a more comprehensive fix tonight for determining when we should quote property names during decompilation. We're still not quoting negative numbers, for instance.

**Comment by bjouhier** [comment]
OK. I'm closing the pull request as you'll handle the fix directly.

**Pull request closed without merging**